_posts/2020-02-05-keycloak-z-angularem-i-keycloakiem.md

```diff
@@ -0,0 +1,427 @@
+        .pipe(
+            filter(config => config.enabled),
+            flatMap(config => {
+                return keycloakService.init({
+                    config: {
+                        url: config.authServerUrl,
+                        realm: config.realm,
+                        clientId: config.resource,
+                        credentials: {
+                            secret: config.credentials.secret
+                        }
+                    },
+                    initOptions: {
+                        onLoad: 'check-sso',
+                        checkLoginIframe: false
+                    }
+                });
+            })).toPromise();
+}
+```
+
+### Zabezpieczenie route'ów
+
+Aby zabezpieczyć routy, wykorzystamy [`Angular Route Guard`](https://angular.io/guide/router#guards).  
+Stworzymy `AppAuthGuard` rozszerzając `KeycloakAuthGuard` oraz implementując [canActivate](https://angular.io/api/router/CanActivate).
+
+W konstruktorze pobierzemy konfigurację z `KeycloakConfigService`, aby zweryfikować czy uwierzytelnianie jest włączone (parametr `keycloak.enabled` w `application.yml`), oraz czy użytkownik posiada wymaganą rolę:
+
+```typescript
+@Injectable({
+    providedIn: 'root'
+})
+export class AppAuthGuard extends KeycloakAuthGuard {
+    isAuthEnabled: boolean = true;
+    requiredUserRole: string;
+
+    constructor(protected router: Router, protected keycloakService: KeycloakService, private keycloakConfigService: KeycloakConfigService) {
+        super(router, keycloakService);
+        this.keycloakConfigService.getConfig().subscribe(config => {
+            this.isAuthEnabled = config.enabled;
+            this.requiredUserRole = config.requiredUserRole;
+        });
+    }
+
+    canActivate(route: ActivatedRouteSnapshot, state: RouterStateSnapshot): Promise<boolean> {
+        if (!this.isAuthEnabled) {
+            return Promise.resolve(true);
+        } else {
+            return super.canActivate(route, state) as Promise<boolean>;
+        }
+    }
+
+    isAccessAllowed(route: ActivatedRouteSnapshot, state: RouterStateSnapshot): Promise<boolean> {
+        return new Promise(async (resolve) => {
+            if (!this.authenticated) {
+                return this.keycloakService.login();
+            }
+            resolve(this.roles.includes(this.requiredUserRole));
+        });
+    }
+}
+```
+
+a następnie wykorzystamy naszego guarda w `app-routing.module.ts` do zabezpieczenia routa `protected`:
+
+```typescript
+const routes: Routes = [
+    {
+        path: '',
+        component: PublicComponent
+    },
+    {
+        path: 'protected',
+        component: ProtectedComponent,
+        canActivate: [AppAuthGuard]
+    }
+];
+```
+
+## Testy
+
+### Uruchomienie
+
+Jeśli mamy już uruchomiony serwer Keycloak, możemy zbudować nasze aplikacje i obraz dockerowy, a następnie uruchomić kontener.
+
+W tym celu w głównym folderze repozytorium najpierw budujemy aplikacje:
```

Comment:
jeżeli tutaj chodzi o jedną aplikację to proszę wstaw "ę" na końcu (nie jestem pewien czy chodzi o jedną czy o kilka).

Reply:
tym poleceniem są budowane dwie aplikacje, backendowa i frontendowa